### ISSUE TITLE
sql: consider `READ COMMITTED` isolation level

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#22574]
## Overview

Consider supporting the [`READ COMMITTED` transaction isolation level](https://www.postgresql.org/docs/current/transaction-iso.html#XACT-READ-COMMITTED).

This is essentially an isolation level that means "ignore my transaction blocks." The data returned by each query in the transaction must execute at a closed timestamp, but there is no requirement that multiple queries in that transaction use the same timestamp.

This isolation level would primarily be useful as an escape hatch when you don't care about consistency and you only care about fast reads, but you can't convince your SQL ORM to stop issuing explicit transactions. (Disabling explicit transactions with SQLAlchemy v2.0 is known to be difficult, for example.)

## Context

A user recently had a performance incident (spiking p95 query latencies) that was caused by a stuck source, despite the transaction isolation level being set to `SERIALIZABLE`. The user was willing to trade staleness for low latency, and expected every query to return immediately even if the source was lagging behind.

The issue was that the source was in the same schema as the views depending on that source and the user's ORM was issuing all `SELECT` statements inside of an explicit transaction:

```
BEGIN;
-- wait for response
SELECT * FROM view;
-- wait for response
COMMIT;
-- wait for response
```

The explicit interactive transaction meant that Materialize needed to pick a timestamp for the `SELECT * FROM view` that was compatible with _all_ object in the view's schema ... including the stuck source.

The user solved this problem by removing the ORM issuing the explicit transactions and replacing it with a simpler PostgreSQL driver. This worked, but was a drastic escape hatch. Another solution would have been to move the source to a new schema, but this would have been a similarly disruptive change. (Also we don't support `ALTER ... SET SCHEMA`.)

Had we supported the `READ COMMITTED` isolation level, we could have instructed the customer to downgrade from the `SERIALIZABLE` isolation level to the `READ COMMITTED` isolation level. This likely would have been an easy change for the customer to apply, and one that they could have applied more quickly.

## Caveats

We might decide that `READ COMMITTED` is off brand. We pride ourselves on our correctness, after all.

On the other hand, we wouldn't made this isolation level the default, and we wouldn't encourage its use except in esoteric situations like the one described above. You could make the case that supporting but discouraging a weak isolation level is still on brand.

**Comment by benesch** [comment]
> > The issue was that the source was in the same schema as the views depending on that source
> > ...
> > The explicit interactive transaction meant that Materialize needed to pick a timestamp for the `SELECT * FROM view` that was compatible with all object in the view's schema ... including the stuck source.
> 
> I'm slightly confused by the context. If the view depended on the source, then wouldn't the view be stuck too? In which case a timestamp compatible with the stuck source is exactly what we want because it will also be compatible with the stuck view?
> 
> I wasn't heavily involved with the recent issue, but my understanding of these classes of issues is that when you select from a stuck view in a multi-statement transaction, then the timestamp needs to be compatible with unstuck objects in the same schema. That's what causes the query on stuck views to hang, it's waiting for the view to get unstuck and catch up with everything else.

That's a very good point. I don't have a super satisfying answer for you. The source has an associated progress collection in the same schema. Perhaps that progress collection continues to advance through time even if the source itself stalls out?

> The only use case I can think of is "you really want serializable single statement transactions, but your ORM/driver keeps converting your single statement into an interactive transaction via `BEGIN` and you're using `READ COMMITTED` as a hacky escape hatch." That feels sort of bad, but I'm not sure what other fix there is. The main alternative seems to be to configure your ORM not to do that, but that's impossible in some cases(?) or at least people don't seem to like to do that.

Yeah, exactly. And probably not something we want you to use long term, but something you can twiddle in the heat of an incident until you can deploy a more permanent fix.

**Comment by jkosh44** [comment]
> > > The issue was that the source was in the same schema as the views depending on that source
> > > ...
> > > The explicit interactive transaction meant that Materialize needed to pick a timestamp for the `SELECT * FROM view` that was compatible with all object in the view's schema ... including the stuck source.
> > 
> > 
> > I'm slightly confused by the context. If the view depended on the source, then wouldn't the view be stuck too? In which case a timestamp compatible with the stuck source is exactly what we want because it will also be compatible with the stuck view?
> > I wasn't heavily involved with the recent issue, but my understanding of these classes of issues is that when you select from a stuck view in a multi-statement transaction, then the timestamp needs to be compatible with unstuck objects in the same schema. That's what causes the query on stuck views to hang, it's waiting for the view to get unstuck and catch up with everything else.
> 
> That's a very good point. I don't have a super satisfying answer for you. The source has an associated progress collection in the same schema. Perhaps that progress collection continues to advance through time even if the source itself stalls out?

That sounds possible. FYI, another alternative that was suggested was to put objects into separate schemas, so they don't affect each others timestamp selection. If progress collections are automatically created in the same schema and they don't get stuck with their associated sources, then the separate schema suggestion doesn't work well.

**Comment by benesch** [comment]
>  If progress collections are automatically created in the same schema and they don't get stuck with their associated sources, then the separate schema suggestion doesn't work well.

Agreed. I think it would have worked okay in this case because the application was only querying the view, and as long as the view was in a different schema from the source and its associated progress collection, all would have been well. But leaves a footgun lying around, assuming that this theory is correct, as an explicit serializable transaction that queries the source directly...

```
BEGIN;
SET transaction_isolation = serializable;
SELECT * FROM source;
COMMIT;
```

...could stall unnecessarily because `source_progress` is racing ahead.
